consensus/src/consensus/inner/commit.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use snarkos_metrics as metrics;
+
+impl ConsensusInner {
+    /// Receive a block from an external source and process it based on ledger state.
+    pub(super) async fn receive_block(&mut self, block: &SerialBlock) -> Result<(), ConsensusError> {
+        let hash = block.header.hash();
+        match self.storage.get_block_state(&hash).await? {
+            BlockStatus::Unknown => (),
+            BlockStatus::Committed(_) | BlockStatus::Uncommitted => {
+                debug!("Received a pre-existing block");
+                metrics::increment_counter!(DUPLICATE_BLOCKS);
+                return Err(ConsensusError::PreExistingBlock);
+            }
+        }
+        self.storage.insert_block(block).await?;
+
+        self.try_commit_block(&hash, block).await?;
+
+        self.try_to_fast_forward().await?;
+
+        Ok(())
+    }
+
+    pub(super) async fn try_commit_block(&mut self, hash: &Digest, block: &SerialBlock) -> Result<(), ConsensusError> {
+        let canon = self.storage.canon().await?;
+
+        match self.storage.get_block_state(&block.header.previous_block_hash).await? {
+            BlockStatus::Committed(n) if n == canon.block_height => {
+                debug!("Processing a block that is on canon chain. Height {} -> {}", n, n + 1);
+                // process now
+            }
+            BlockStatus::Unknown => {
+                debug!("Processing a block that is an unknown orphan");
+
+                // dont process
+                return Ok(());
+            }
+            _ => {
```

Comment:
I think it would be a readability win to put the body of this branch in a new method, e.g. `handle_potential_fork`